### PR TITLE
Dom0 services

### PR DESCRIPTION
**Pull request by fepitre**

### CODE REVIEW

File: qubes/config.py [hunk 1]
```diff
@@ -44,6 +44,8 @@
     # qubes_icon_dir is obsolete
     # use QIcon.fromTheme() where applicable
     'qubes_icon_dir': '/usr/share/icons/hicolor/128x128/devices',
+
+    'dom0_services_dir': '/var/lib/qubes-services',
```
Comment (marmarek) [inline qubes/config.py]: This should be `/var/run/qubes-service`.

Comment (fepitre) [reply qubes/config.py]: Yes indeed.

File: qubes/config.py [hunk 2]
```diff
@@ -44,6 +44,8 @@
     # qubes_icon_dir is obsolete
     # use QIcon.fromTheme() where applicable
     'qubes_icon_dir': '/usr/share/icons/hicolor/128x128/devices',
+
+    'dom0_services_dir': '/var/run/qubes-services',
```
Comment (marmarek) [inline qubes/config.py]: It is `/var/run/qubes-service` (without `s` at the end).

File: qubes/ext/services.py [hunk 3]
```diff
@@ -95,9 +129,18 @@ def on_domain_load(self, vm, event):
...
             del vm.features['service.meminfo-writer']
 
+        if vm.name == "dom0":
+            for feature, value in vm.features.items():
+                if not feature.startswith('service.'):
+                    continue
+                service = feature[len('service.'):]
+                if str(int(bool(value))) == "1" and not os.path.exists(
+                        '/var/run/qubes-service/{}'.format(service)):
+                    os.mknod('/var/run/qubes-service/{}'.format(service))
```
Comment (marmarek) [inline qubes/ext/services.py]: This should use the `add_dom0_services` function.

File: qubes/ext/services.py [hunk 4]
```diff
@@ -18,34 +18,62 @@
...
 
 class ServicesExtension(qubes.ext.Extension):
-    '''This extension export features with 'service.' prefix to QubesDB in
+    """This extension export features with 'service.' prefix to QubesDB in
     /qubes-service/ tree.
-    '''
+    """
+
+    @staticmethod
+    def add_dom0_services(vm, service):
```
Comment (marmarek) [inline qubes/ext/services.py]: Since it takes just one service, I'd use singular `service` in the function name.

File: qubes/ext/services.py [hunk 5]
```diff
@@ -18,34 +18,62 @@
...
+            service = '{}/{}'.format(
+                qubes.config.system_path['dom0_services_dir'], service)
+            if not os.path.exists(service):
+                os.mknod(service)
+        except PermissionError:
+            vm.log.warning("Cannot write to {}".format(
+                qubes.config.system_path['dom0_services_dir']))
+
+    @staticmethod
+    def remove_dom0_services(vm, service):
```
Comment (marmarek) [inline qubes/ext/services.py]: And same here: `remove_dom0_service`

File: qubes/tests/ext.py [hunk 6]
```diff
@@ -322,3 +330,38 @@ def test_012_supported_services_remove(self):
...
+            'feature-set:service.service.guivm-gui-agent',
+            'service.guivm-gui-agent', '1')
+
+        self.assertEqual(os.path.exists(service_path), True)
+
+    def test_014_feature_delete_dom0(self):
+        self.test_base_dir = '/tmp/qubes-test-dir'
+        self.base_dir_patch = mock.patch.dict(
+            qubes.config.system_path, {'dom0_services_dir': self.test_base_dir})
+        self.base_dir_patch.start()
```
Comment (marmarek) [inline qubes/tests/ext.py]: You need to stop the patch somewhere. For example add `self.addCleanup(self.base_dir_patch.stop)`

File: qubes/tests/ext.py [hunk 7]
```diff
@@ -322,3 +330,38 @@ def test_012_supported_services_remove(self):
         self.assertEqual(self.features, {
             'supported-service.test2': True,
         })
+
+    def test_013_feature_set_dom0(self):
+        self.test_base_dir = '/tmp/qubes-test-dir'
+        self.base_dir_patch = mock.patch.dict(
+            qubes.config.system_path, {'dom0_services_dir': self.test_base_dir})
+        self.base_dir_patch.start()
```
Comment (marmarek) [inline qubes/tests/ext.py]: You need to stop the patch somewhere. For example add `self.addCleanup(self.base_dir_patch.stop)`

File: qubes/tests/ext.py [hunk 8]
```diff
@@ -322,3 +330,38 @@ def test_012_supported_services_remove(self):
...
+
+        self.ext.on_domain_feature_set(
+            self.dom0,
+            'feature-set:service.service.guivm-gui-agent',
+            'service.guivm-gui-agent', '1')
+
+        self.ext.on_domain_feature_delete(
+            self.dom0,
+            'feature-delete:service.service.guivm-gui-agent',
+            'service.guivm-gui-agent')
```
Comment (marmarek) [inline qubes/tests/ext.py]: Check also if setting empty value removes the service too.

File: qubes/tests/ext.py [hunk 9]
```diff
@@ -322,3 +330,53 @@ def test_012_supported_services_remove(self):
...
+            qubes.config.system_path, {'dom0_services_dir': self.test_base_dir})
+        self.base_dir_patch.start()
+        service = 'guivm-gui-agent'
+        service_path = self.test_base_dir + '/' + service
+
+        self.ext.on_domain_feature_set(
+            self.dom0,
+            'feature-set:service.service.guivm-gui-agent',
+            'service.guivm-gui-agent', '1')
+        self.addCleanup(self.base_dir_patch.stop)
```
Comment (marmarek) [inline qubes/tests/ext.py]: Better place addCleanup directly below the code it cleanup after.

File: qubes/tests/ext.py (hunk 6 above)
Comment (marmarek) [reply qubes/tests/ext.py]: This still applies.

File: qubes/tests/ext.py [hunk 10]
```diff
@@ -322,3 +330,53 @@ def test_012_supported_services_remove(self):
...
+            qubes.config.system_path, {'dom0_services_dir': self.test_base_dir})
+        self.base_dir_patch.start()
+        service = 'guivm-gui-agent'
+        service_path = self.test_base_dir + '/' + service
+
+        self.ext.on_domain_feature_set(
+            self.dom0,
+            'feature-set:service.service.guivm-gui-agent',
+            'service.guivm-gui-agent', '')
+        self.addCleanup(self.base_dir_patch.stop)
```
Comment (marmarek) [inline qubes/tests/ext.py]: And here too - should be just below patch.start()

**Pull request merged**